### PR TITLE
Reduce iterations required for dead argument elimination convergence

### PR DESCRIPTION
**Pull request by arsnyder16**
Reduce iterations required for dead argument elimination convergence

Crude measurements 
Before:
```
[PassRunner]   running pass: dae-optimizing...                     53.1484 seconds.
```
After:
```
[PassRunner]   running pass: dae-optimizing...                     39.9249 seconds.
```

<details>
<summary>More details</summary>
<details>
<summary>Before</summary>

```
[PassRunner] running passes
[PassRunner]   running pass: dae-optimizing...                     
Function Count: 132035

Iteration: 1
  removedParams: 12961
  returnValuesRemoved: 0
Iteration: 2
  removedParams: 2343
  returnValuesRemoved: 0
Iteration: 3
  removedParams: 903
  returnValuesRemoved: 0
Iteration: 4
  removedParams: 553
  returnValuesRemoved: 0
Iteration: 5
  removedParams: 409
  returnValuesRemoved: 0
Iteration: 6
  removedParams: 303
  returnValuesRemoved: 0
Iteration: 7
  removedParams: 124
  returnValuesRemoved: 0
Iteration: 8
  removedParams: 100
  returnValuesRemoved: 0
Iteration: 9
  removedParams: 88
  returnValuesRemoved: 0
Iteration: 10
  removedParams: 69
  returnValuesRemoved: 0
Iteration: 11
  removedParams: 53
  returnValuesRemoved: 0
Iteration: 12
  removedParams: 41
  returnValuesRemoved: 0
Iteration: 13
  removedParams: 21
  returnValuesRemoved: 0
Iteration: 14
  removedParams: 17
  returnValuesRemoved: 0
Iteration: 15
  removedParams: 12
  returnValuesRemoved: 0
Iteration: 16
  removedParams: 10
  returnValuesRemoved: 0
Iteration: 17
  removedParams: 7
  returnValuesRemoved: 0
Iteration: 18
  removedParams: 4
  returnValuesRemoved: 0
Iteration: 19
  removedParams: 4
  returnValuesRemoved: 0
Iteration: 20
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 21
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 22
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 23
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 24
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 25
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 26
  removedParams: 0
  returnValuesRemoved: 16824
Iteration: 27
  removedParams: 11
  returnValuesRemoved: 0
Iteration: 28
  removedParams: 1
  returnValuesRemoved: 0
Iteration: 29
  removedParams: 0
  returnValuesRemoved: 1523
Iteration: 30
  removedParams: 1
  returnValuesRemoved: 0
Iteration: 31
  removedParams: 0
  returnValuesRemoved: 101
Iteration: 32
  removedParams: 0
  returnValuesRemoved: 73
Iteration: 33
  removedParams: 0
  returnValuesRemoved: 24
Iteration: 34
  removedParams: 0
  returnValuesRemoved: 2
Iteration: 35
  removedParams: 0
  returnValuesRemoved: 1

Total Iterations: 35

Total removedParams: 18047

Total returnValuesRemoved : 18548
53.1484 seconds.
```

</details>
<details>
<summary>After</summary>

```
[PassRunner] running passes
[PassRunner]   running pass: dae-optimizing...                     
Function Count: 132035

Iteration: 1
  removedParams: 12961
  returnValuesRemoved: 15818
Iteration: 2
  removedParams: 2352
  returnValuesRemoved: 2472
Iteration: 3
  removedParams: 906
  returnValuesRemoved: 154
Iteration: 4
  removedParams: 553
  returnValuesRemoved: 78
Iteration: 5
  removedParams: 409
  returnValuesRemoved: 27
Iteration: 6
  removedParams: 303
  returnValuesRemoved: 2
Iteration: 7
  removedParams: 124
  returnValuesRemoved: 1
Iteration: 8
  removedParams: 100
  returnValuesRemoved: 0
Iteration: 9
  removedParams: 88
  returnValuesRemoved: 0
Iteration: 10
  removedParams: 69
  returnValuesRemoved: 0
Iteration: 11
  removedParams: 53
  returnValuesRemoved: 0
Iteration: 12
  removedParams: 41
  returnValuesRemoved: 0
Iteration: 13
  removedParams: 21
  returnValuesRemoved: 0
Iteration: 14
  removedParams: 17
  returnValuesRemoved: 0
Iteration: 15
  removedParams: 12
  returnValuesRemoved: 0
Iteration: 16
  removedParams: 10
  returnValuesRemoved: 0
Iteration: 17
  removedParams: 7
  returnValuesRemoved: 0
Iteration: 18
  removedParams: 4
  returnValuesRemoved: 0
Iteration: 19
  removedParams: 4
  returnValuesRemoved: 0
Iteration: 20
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 21
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 22
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 23
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 24
  removedParams: 2
  returnValuesRemoved: 0
Iteration: 25
  removedParams: 2
  returnValuesRemoved: 0

Total Iterations: 25

Total removedParams: 18046

Total returnValuesRemoved : 18552
39.9249 seconds.
```

</details>

</details>

### CODE REVIEW

File: src/passes/DeadArgumentElimination.cpp [hunk 1]
```diff
@@ -105,7 +104,7 @@ struct DAEScanner
     }
     if (curr->isReturn) {
       info->hasTailCalls = true;
-      info->tailCallees.insert(curr->target);
+      (*infoMap)[curr->target].hasTailCalls = true;
```
Comment (kripken) [inline src/passes/DeadArgumentElimination.cpp:107]: Should the field `hasTailCalls` be atomic for this to work? (we do that for `hasUnseenCalls`)

**Review by kripken — COMMENTED** [review]
Nice! Looks right to me aside from that comment.

Please fuzz this locally (`./scripts/fuzz_opt.py`). Just FuzzExec is enough here, so you can apply this diff locally before fuzzing to avoid needing to install d8 etc.,
```diff
diff --git a/scripts/fuzz_opt.py b/scripts/fuzz_opt.py
index 5b6e9799e..ffeb54c8d 100755
--- a/scripts/fuzz_opt.py
+++ b/scripts/fuzz_opt.py
@@ -1012,12 +1012,6 @@ class RoundtripText(TestCaseHandler):
 # The global list of all test case handlers
 testcase_handlers = [
     FuzzExec(),
-    CompareVMs(),
-    CheckDeterminism(),
-    Wasm2JS(),
-    Asyncify(),
-    # FIXME: Re-enable after https://github.com/WebAssembly/binaryen/issues/3989
-    # RoundtripText()
 ]
 
 
```

File: src/passes/DeadArgumentElimination.cpp (hunk 1 above)
Comment (arsnyder16) [reply src/passes/DeadArgumentElimination.cpp:107]: yes! good call, for some reason i thought i did that, checked my original PR and missed it there as well 👎

Comment (arsnyder16) [reply src/passes/DeadArgumentElimination.cpp:107]: @kripken Will `fuzz_opt.py` finish on it own or should i leave it run for a certain amount of time?

Comment (kripken) [reply src/passes/DeadArgumentElimination.cpp:107]: It runs forever. If you see it run for several thousand iterations, that's probably enough here.

Comment (arsnyder16) [reply src/passes/DeadArgumentElimination.cpp:110]: @kripken Well i was away for a few hours and it ran 15k iterations but it did find something. Not sure if its related to these changes or not 

<details>
<summary>Details</summary>

```

You found a bug! Please report it with

  seed: 16576256188251122925

and the exact version of Binaryen you found it on, plus the exact Python
version (hopefully deterministic random numbers will be identical).

You can run that testcase again with "fuzz_opt.py 16576256188251122925"

The initial wasm file used here is saved as /root/binaryen/out/test/original.wasm

You can reduce the testcase by running this now:

||||
vvvv


/root/binaryen/build-clang/bin/wasm-reduce --nominal /root/binaryen/out/test/original.wasm '--command=bash /root/binaryen/out/test/reduce.sh' -t /root/binaryen/out/test/t.wasm -w /root/binaryen/out/test/w.wasm


^^^^
||||

Make sure to verify by eye that the output says something like this:

At least one of the next two values should be 0:
  0
  1
The following value should be 1:
  1

(If it does not, then one possible issue is that the fuzzer fails to write a
valid binary. If so, you can print the output of the fuzzer's first command
(using -ttf / --translate-to-fuzz) in text form and run the reduction from that,
passing --text to the reducer.)

You can also read "/root/binaryen/out/test/reduce.sh" which has been filled out for you and includes
docs and suggestions.

After reduction, the reduced file will be in /root/binaryen/out/test/w.wasm
================================================================================
```
<details>

Comment (kripken) [reply src/passes/DeadArgumentElimination.cpp:110]: Hmm, I don't see an error locally with that seed. Can you paste the full output of just running that one seed, `./scripts/fuzz_opt.py 16576256188251122925` ?

Comment (arsnyder16) [reply src/passes/DeadArgumentElimination.cpp:110]: Ok its passing for me, its possible i had binaries from my follow up PR and maybe there is an issue you there. I will check it that before i post it. 

Just ran the fuzz_opt.py again for 3500 integrations and no issues

**Comment by sbc100** [comment]
This change is currently breaking the binaryen roll into emscripten.

I bisected the test failure to this PR:

```
$ ./tests/runner cores.test_getcwd_with_non_ascii_name
....
RuntimeError: Aborted(Assertion failed: longStr.length() == 255, at: /usr/local/google/home/sbc/dev/wasm/emscripten/tests/fs/test_getcwd_with_non_ascii_name.cpp,70,main). Build with -sASSERTIONS for more info.
    at abort (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:7906)
    at ___assert_fail (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:11285)
    at k (<anonymous>:wasm-function[52]:0x3aa0)
    at Module._main (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60302)
    at callMain (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60688)
    at doRun (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:61107)
    at run (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:61279)
    at runCaller (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60543)
    at removeRunDependency (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:7710)
    at receiveInstance (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:9420)
    at receiveInstantiationResult (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:9541)
Thrown at:
    at abort (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:7906)
    at ___assert_fail (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:11285)
    at k (:1:15009)
    at Module._main (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60302)
    at callMain (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60688)
    at doRun (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:61107)
    at run (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:61279)
    at runCaller (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:60543)
    at removeRunDependency (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:7710)
    at receiveInstance (/tmp/emscripten_test_cores_xgqdan7k/test_getcwd_with_non_ascii_name.js:1:9420)
```

Only occurs with `-Os` not `-O2`.  Seems to change program semantics in such as way that the assertion in the test program fails.

**Comment by arsnyder16** [comment]
@sbc100 @kripken #4646

**Comment by sbc100** [comment]
If you would prefer to fix forward rather than revert the current breakage is not actually hurting emscripten yet, its just preventing new changes from binaryen making their way to emscripten users.   Just FYI.

**Comment by arsnyder16** [comment]
@sbc100 I am struggling to reproduce. 
Here is what i am trying.  `test.baseline.wasm` and  `test.new.wasm` are binary matches 
```bash

em++ ../../emscripten/tests/fs/test_getcwd_with_non_ascii_name.cpp -O0 -o test.js
node test.js
mv test.wasm test.og.wasm

/root/emsdk/upstream/bin/wasm-opt -Os \
 --strip-dwarf --post-emscripten --low-memory-unused --zero-filled-memory --strip-debug --strip-producers \
 test.og.wasm -o test.wasm --mvp-features
node test.js
mv test.wasm test.baseline.wasm

bin/wasm-opt -Os \
 --strip-dwarf --post-emscripten --low-memory-unused --zero-filled-memory --strip-debug --strip-producers \
 test.og.wasm -o test.wasm --mvp-features
node test.js
mv test.wasm test.new.wasm

```

```
~/binaryen/build (asnyder/dae-iterations2)# ./test.sh
positive test case: абвгд
ret2 = /абвгд
positive test case: abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
ret2 = /abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
negative test case: abcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcde
positive test case: абвгд
ret2 = /абвгд
positive test case: abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
ret2 = /abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
negative test case: abcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcde
positive test case: абвгд
ret2 = /абвгд
positive test case: abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
ret2 = /abcdeабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгдабвгд
negative test case: abcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcdeabcde
```

**Comment by sbc100** [comment]
Did you try running exact test command from above `./tests/runner cores.test_getcwd_with_non_ascii_name` ?

**Comment by arsnyder16** [comment]
I don't have the full emscripten dev environment setup, also with it so i can integrate local binaryen.  

But i now am able to repro but swapping my local build of wasm-opt into my emsdk upstream then just running

```
em++ ../emscripten/tests/fs/test_getcwd_with_non_ascii_name.cpp -Os -o test.js
node test.js
```

Thanks! I'll get to the bottom of this

**Comment by sbc100** [comment]
emscripten ships with the test runner and test suite.   Just run `./tests/runner` from either and emscirpten checkout or the emskd version of emscipten (which em++ to see where that is).

You can configure emscripten to use your custom binaryen version using `EM_BINARYEN_ROOT` in your environment.

**Comment by arsnyder16** [comment]
awesome! thanks!

**Pull request merged**